identity-account/src/account/account.rs

```diff
@@ -255,166 +252,151 @@ impl Account {
       .await?;
 
-    let message: MessageId = if self.config.testmode {
-      MessageId::null()
+    log::debug!(
+      "[publish_integration_change] publishing on index {}",
+      new_doc.integration_index()
+    );
+
+    let message_id: MessageId = if self.config.testmode {
+      // Fake publishing by returning a random message id.
+      MessageId::new(unsafe { crypto::utils::rand::gen::<[u8; 32]>().unwrap() })
     } else {
-      self
-        .client_map
-        .publish_diff(old_state.this_message_id(), &diff)
-        .await?
-        .into()
+      self.client_map.publish_document(&new_doc).await?.into()
     };
 
-    let events: [Event; 1] = [Event::new(EventData::DiffMessage(message))];
-
-    Self::commit_events(self.did(), &self.config, self.storage(), &new_root, &events).await?;
+    self.chain_state.set_last_integration_message_id(message_id);
 
     Ok(())
   }
 
-  async fn fold_snapshot(snapshot: IdentitySnapshot, commit: Commit) -> Result<IdentitySnapshot> {
-    Ok(IdentitySnapshot {
-      sequence: commit.sequence().max(snapshot.sequence),
-      identity: commit.into_event().apply(snapshot.identity).await?,
-    })
-  }
-
-  #[doc(hidden)]
-  pub async fn load_snapshot(&self) -> Result<IdentitySnapshot> {
-    // Retrieve the state snapshot from storage or create a new one.
-    let initial: IdentitySnapshot = self
-      .storage()
-      .snapshot(self.did())
-      .await?
-      .unwrap_or_else(|| IdentitySnapshot::new(IdentityState::new()));
-
-    // Apply all recent events to the state and create a new snapshot
-    self
-      .storage()
-      .stream(self.did(), initial.sequence())
-      .await?
-      .try_fold(initial, Self::fold_snapshot)
-      .await
-  }
+  async fn publish_diff_change(&mut self, old_state: &IdentityState) -> Result<()> {
+    log::debug!("[publish_diff_change] publishing {:?}", self.document().did());
 
-  async fn load_snapshot_at(&self, generation: Generation) -> Result<IdentitySnapshot> {
-    let initial: IdentitySnapshot = IdentitySnapshot::new(IdentityState::new());
+    let old_doc: &IotaDocument = old_state.document();
+    let new_doc: &IotaDocument = self.state().document();
 
-    // Apply all events up to `generation`
-    self
-      .storage()
-      .stream(self.did(), Generation::new())
-      .await?
-      .take(generation.to_u32() as usize)
-      .try_fold(initial, Self::fold_snapshot)
-      .await
-  }
+    let mut diff_id: &MessageId = self.chain_state().last_diff_message_id();
 
-  async fn commit_events(
-    did: &IotaDID,
-    config: &AccountConfig,
-    storage: &dyn Storage,
-    state: &IdentitySnapshot,
-    events: &[Event],
-  ) -> Result<Vec<Commit>> {
-    // Bail early if there are no new events
-    if events.is_empty() {
-      return Ok(Vec::new());
+    // If there was no previous diff message, use the previous int message.
+    if diff_id.is_null() {
+      diff_id = self.chain_state.last_integration_message_id();
```

Comment:
Maybe we should return an error after this if the `diff_id` is still `null`? It would mean something went wrong with the chain state but it would avoid publishing an orphaned diff update with no chain.

Reply:
We can check, but the `diff_id` can only be null here when storing the chain state after the last update failed or the chain state got corrupted in some other way, in which case I would like to return a fatal error. When it happens, it's not the user's fault, so a fatal error seems appropriate. I would like to avoid adding another edge case error type, as that will just make life harder for the error refactoring, and instead leave a TODO there.

Reply:
I'm not sure about the error variants but I do think we need something to indicate that the diff update will be invalid. A TODO should be fine for now.